src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs

```diff
@@ -83,32 +83,101 @@ internal partial class CodeFixService : ForegroundThreadAffinitizedObject, ICode
 
             using (var diagnostics = SharedPools.Default<List<DiagnosticData>>().GetPooledObject())
             {
-                var fullResult = await _diagnosticService.TryAppendDiagnosticsForSpanAsync(document, range, diagnostics.Object, cancellationToken: cancellationToken).ConfigureAwait(false);
-                foreach (var diagnostic in diagnostics.Object)
-                {
-                    cancellationToken.ThrowIfCancellationRequested();
+                var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                var linkedToken = linkedTokenSource.Token;
 
-                    if (!range.IntersectsWith(diagnostic.TextSpan))
-                    {
-                        continue;
-                    }
+                var fullResult = await _diagnosticService.TryAppendDiagnosticsForSpanAsync(document, range, diagnostics.Object, cancellationToken: linkedToken).ConfigureAwait(false);
+
+                var diagnosticGroups = diagnostics.Object.GroupBy(d => d.Severity);
+                var nonErrorDiagnosticGroups = diagnosticGroups.Where(d => d.Key != DiagnosticSeverity.Error);
+                var taskSet = nonErrorDiagnosticGroups.Select(severityGroup =>
```

Comment:
Though I just realized that since I'm treating all non-error diagnostics the same I can just kick off one task that processes all of the non-error diagnostics (rather than 3 of them).